### PR TITLE
#35 Setup Coverage Reports

### PR DESCRIPTION
**Pull request by alchuu00**
- decided to remain using Istanbul package (https://istanbul.js.org/) for code coverage reports since it already comes with Jest
- created npm command to display code coverage report

I am not sure which folders and files should I exclude from the report?
Also I am not sure if treshold 80% is okay? Many articles state 80% is good to aim for.

### CODE REVIEW

**Comment by umsu2** [comment]
i think i mentioned before the test coverage is already setup. it was set at 30pct. i do like we are setting it to 80pct if possible.

File: package.json [hunk 1]
```diff
@@ -38,5 +39,12 @@
     "jest": "^29.7.0",
     "jest-environment-jsdom": "^29.7.0",
     "typescript": "5.2.2"
+  },
```
Comment (umsu2) [inline package.json:42]: i dont think this is needed. if you want to change the coverage path in the above file

File: package.json [hunk 2]
```diff
@@ -9,7 +9,8 @@
     "lint:check": "next lint",
     "startProduction": "next start",
     "test": "jest",
-    "testWatch": "jest --watch"
+    "testWatch": "jest --watch",
+    "test:coverage": "jest --coverage --config jest.config.mjs"
```
Comment (umsu2) [inline package.json:13]: dont need this.

File: jest.config.mjs [hunk 3]
```diff
@@ -14,11 +14,19 @@ const config = {
...
+    '!**/__tests__/__snapshots__/**',
+    '!**/enum/**',
+  ],
   coverageThreshold: {
     global: {
-      lines: 30
-    }
-  }
+      statements: 80,
+      branches: 80,
```
Comment (umsu2) [inline jest.config.mjs:25]: :+1:

**Comment by alchuu00** [comment]
@umsu2 I made the changes you suggested to `package.json`.

**Comment by alchuu00** [comment]
Also this is the test output now:
```
---------------------------|---------|----------|---------|---------|-------------------
File                       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
---------------------------|---------|----------|---------|---------|-------------------
All files                  |   59.59 |     62.5 |    62.5 |   59.57 |                   
 app                       |       0 |      100 |       0 |       0 |                   
  layout.tsx               |       0 |      100 |       0 |       0 | 1-9               
  page.tsx                 |       0 |      100 |       0 |       0 | 2-10              
 app/_component            |       0 |      100 |       0 |       0 | 
  home-page.tsx            |       0 |      100 |       0 |       0 | 1-3
 app/about                 |     100 |      100 |     100 |     100 | 
  page.tsx                 |     100 |      100 |     100 |     100 | 
 app/join-us               |     100 |      100 |     100 |     100 | 
  page.tsx                 |     100 |      100 |     100 |     100 | 
 app/projects              |       0 |      100 |       0 |       0 | 
  page.tsx                 |       0 |      100 |       0 |       0 | 3-29              
 client-services           |     100 |       50 |     100 |     100 | 
  github.client-service.ts |     100 |       50 |     100 |     100 | 29
 components/banner         |     100 |      100 |     100 |     100 | 
  banner.tsx               |     100 |      100 |     100 |     100 | 
 components/footer         |       0 |      100 |       0 |       0 | 
  footer.tsx               |       0 |      100 |       0 |       0 | 3-21
 components/header         |      95 |      100 |      75 |   94.44 | 
  header-data.tsx          |     100 |      100 |     100 |     100 | 
  header.tsx               |   93.75 |      100 |      75 |   93.33 | 26
 components/project        |       0 |        0 |       0 |       0 | 
  project.tsx              |       0 |        0 |       0 |       0 | 3-27              
 mapper                    |     100 |      100 |     100 |     100 | 
  github.mapping.ts        |     100 |      100 |     100 |     100 | 
 utils/header-map          |     100 |      100 |     100 |     100 | 
  header-map.tsx           |     100 |      100 |     100 |     100 | 
---------------------------|---------|----------|---------|---------|-------------------
Jest: "global" coverage threshold for statements (80%) not met: 59.59%
Jest: "global" coverage threshold for branches (80%) not met: 62.5%
Jest: "global" coverage threshold for lines (80%) not met: 59.57%
Jest: "global" coverage threshold for functions (80%) not met: 62.5%
```

Based on this we need to write additional unit tests for the files above that don't have sufficient coverage.
Should I open a new issue and start on that or do I have to make aditional changes to this PR?

**Comment by SamHuo213** [comment]
> Also this is the test output now:
> 
> ```
> ---------------------------|---------|----------|---------|---------|-------------------
> File                       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
> ---------------------------|---------|----------|---------|---------|-------------------
> All files                  |   59.59 |     62.5 |    62.5 |   59.57 |                   
>  app                       |       0 |      100 |       0 |       0 |                   
>   layout.tsx               |       0 |      100 |       0 |       0 | 1-9               
>   page.tsx                 |       0 |      100 |       0 |       0 | 2-10              
>  app/_component            |       0 |      100 |       0 |       0 | 
>   home-page.tsx            |       0 |      100 |       0 |       0 | 1-3
>  app/about                 |     100 |      100 |     100 |     100 | 
>   page.tsx                 |     100 |      100 |     100 |     100 | 
>  app/join-us               |     100 |      100 |     100 |     100 | 
>   page.tsx                 |     100 |      100 |     100 |     100 | 
>  app/projects              |       0 |      100 |       0 |       0 | 
>   page.tsx                 |       0 |      100 |       0 |       0 | 3-29              
>  client-services           |     100 |       50 |     100 |     100 | 
>   github.client-service.ts |     100 |       50 |     100 |     100 | 29
>  components/banner         |     100 |      100 |     100 |     100 | 
>   banner.tsx               |     100 |      100 |     100 |     100 | 
>  components/footer         |       0 |      100 |       0 |       0 | 
>   footer.tsx               |       0 |      100 |       0 |       0 | 3-21
>  components/header         |      95 |      100 |      75 |   94.44 | 
>   header-data.tsx          |     100 |      100 |     100 |     100 | 
>   header.tsx               |   93.75 |      100 |      75 |   93.33 | 26
>  components/project        |       0 |        0 |       0 |       0 | 
>   project.tsx              |       0 |        0 |       0 |       0 | 3-27              
>  mapper                    |     100 |      100 |     100 |     100 | 
>   github.mapping.ts        |     100 |      100 |     100 |     100 | 
>  utils/header-map          |     100 |      100 |     100 |     100 | 
>   header-map.tsx           |     100 |      100 |     100 |     100 | 
> ---------------------------|---------|----------|---------|---------|-------------------
> Jest: "global" coverage threshold for statements (80%) not met: 59.59%
> Jest: "global" coverage threshold for branches (80%) not met: 62.5%
> Jest: "global" coverage threshold for lines (80%) not met: 59.57%
> Jest: "global" coverage threshold for functions (80%) not met: 62.5%
> ```
> 
> Based on this we need to write additional unit tests for the files above that don't have sufficient coverage. Should I open a new issue and start on that or do I have to make aditional changes to this PR?

Is there coverage reports that show line by line?

**Comment by alchuu00** [comment]
@SamHuo213 I am not sure if I understand what you mean. 

Report like this?


Screenshot_3

**Comment by SamHuo213** [comment]
> @SamHuo213 I am not sure if I understand what you mean.
> 
> Report like this?
> 
> Screenshot_3

yup, does that get generated?

**Comment by SamHuo213** [comment]
Another question, since our coverage does not meet the 80% threshold. What happens?

**Comment by alchuu00** [comment]
> Another question, since our coverage does not meet the 80% threshold. What happens?

I believe that if our coverage doesn't meet 80% threshold then whenever we try to make pull request merge it will fail.

**Comment by SamHuo213** [comment]
> > Another question, since our coverage does not meet the 80% threshold. What happens?
> 
> I believe that if our coverage doesn't meet 80% threshold then whenever we try to make pull request merge it will fail.

That's weird, I don't see any checks.

1) I think we need to ensure that if the coverage is not above 80% it fails the pipeline.
2) I think we need a plan to get our coverage up to 80%

2.1) Either leave it at 30% now and create ticket(s) to add more test and increase it to 80%
OR
2.2) Increase it to 80%, let the pipeline give warning. Add tickets to increase it to 80% then change it to fail.

cc: @umsu2

**Comment by umsu2** [comment]
> > > Another question, since our coverage does not meet the 80% threshold. What happens?
> > 
> > 
> > I believe that if our coverage doesn't meet 80% threshold then whenever we try to make pull request merge it will fail.
> 
> That's weird, I don't see any checks.
> 
> 1. I think we need to ensure that if the coverage is not above 80% it fails the pipeline.
> 2. I think we need a plan to get our coverage up to 80%
> 
> 2.1) Either leave it at 30% now and create ticket(s) to add more test and increase it to 80% OR 2.2) Increase it to 80%, let the pipeline give warning. Add tickets to increase it to 80% then change it to fail.
> 
> cc: @umsu2

no checks because the PR is not up to date with master @alchuu00 let me know when you have time we can go over resolving git conflicts.

**Comment by umsu2** [comment]
and in terms of coverage report. it does get generated automatically when we run tests. BUT the coverage report (html) gets thrown away unless we store/host them somewhere.

File: jest.config.mjs [hunk 4]
```diff
@@ -1,5 +1,10 @@
 import nextJest from 'next/jest.js';
 
+const date = new Date();
+const timestamp = `${date.getFullYear()}-${
+  date.getMonth() + 1
+}-${date.getDate()}-${date.getHours()}-${date.getMinutes()}`;
+
```
Comment (umsu2) [inline jest.config.mjs:7]: not sure what the intention is for having date in the coverage report. 
we don't commit coverage report in our code base.

Comment (umsu2) [reply jest.config.mjs:7]: everything else looks good.

Comment (alchuu00) [reply jest.config.mjs:7]: > not sure what the intention is for having date in the coverage report. we don't commit coverage report in our code base.

So we don't need to save every coverage report? Do we just need to save the current/most recent one which gets overwritten with new one when we run tests again?

**Comment by alchuu00** [comment]
@umsu2 I removed date from the coverage report filename

**Comment by umsu2** [comment]
looks good, just need to update your branch with main again.

**Review by umsu2 — APPROVED** [review]

**Pull request merged**